File: apple/Sources/ReactNativeWatchOS/Modifiers/SwipeActions.swift
import SwiftUI

/// SwiftUI `.swipeActions(edge:allowsFullSwipe:) { … }`. Renders the content
/// slot (the action `Button`s) as swipe actions on a list row. `edge` maps
/// 'leading'/'trailing' to `HorizontalEdge` (default `.trailing`);
/// `allowsFullSwipe` defaults to true. No-ops when no content was provided.
enum RNWSwipeActionsModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("swipeActions") { view, params, ctx in
            guard let body = ctx.content(params.string("content")) else { return view }
            let edge: HorizontalEdge = params.string("edge") == "leading" ? .leading : .trailing
            let allowsFullSwipe = params.bool("allowsFullSwipe") ?? true
            return AnyView(view.swipeActions(edge: edge, allowsFullSwipe: allowsFullSwipe) {
                body
            })
        }
    }
}

File: src/modifiers/swipeActions.ts
import type { ReactNode } from 'react';

import { createModifier } from './createModifier';

export type SwipeEdge = 'leading' | 'trailing';

export interface SwipeActionsParams {
  /// Which edge the actions are revealed from. Defaults to `'trailing'`.
  edge?: SwipeEdge;
  /// Whether a full swipe triggers the first action. Defaults to true.
  allowsFullSwipe?: boolean;
  /// The swipe action buttons. Hoisted into a content slot by `useModifiers`.
  content: ReactNode;
}

/// SwiftUI `.swipeActions(edge:allowsFullSwipe:) { … }`. Attaches swipe
/// actions to a list row; `content` holds the action `Button`s.
///
/// ```tsx
/// <Text modifiers={[swipeActions({ content: <Button title="Delete" .../> })]} />
/// ```
export function swipeActions(content: ReactNode): ReturnType<typeof createModifier>;
export function swipeActions(
  params: SwipeActionsParams
): ReturnType<typeof createModifier>;
export function swipeActions(a: SwipeActionsParams | ReactNode) {
  if (a != null && typeof a === 'object' && 'content' in a) {
    const p = a as SwipeActionsParams;
    return createModifier('swipeActions', {
      edge: p.edge,
      allowsFullSwipe: p.allowsFullSwipe,
      content: p.content,
    });
  }
  return createModifier('swipeActions', { content: a as ReactNode });
}
